[ZIP 203] Transaction Expiry

arcalinea opened the pull request.

str4d commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,61 @@
+Specification
+===============
+
+Transactions will have a new field, nBlockExpiry, which will set the blockheight after which transactions will be removed from the mempool if they have not been mined. The last block that the transaction below could possibly be included in is 3539. After that, it will be removed from the mempool.
```
Is there any benefit to making this `nExpiryTime`, and allowing it to be either a time or a block height, with same semantics as `nLockTime`? How well do Bitcoin users reason about `nLockTime`?

arcalinea replied:
Yeah, `nLockTime`,  is commonly used by `OP_CHECKLOCKTIMEVERIFY` scripts, think it would make sense to follow the semantics as `nExpiryTime`.

As for using both `nLockTime` and `nExpiryTime` together, I could imagine a use case where a time range for when a transaction could be committed to the chain is set. If participants are passing transactions to record state off-chain, this could allow transactions to expire if they're not published and mined within a certain time period.

str4d commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,61 @@
+
+sendtoaddress will allow the user to easily set a blockheight for the transaction to expire.
+
+sendtoaddress "zcashaddress" amount ( blockexpiry "comment" "comment-to" subtractfeefromamount )
```
Hmm... We (and other implementers) will have to take care here, given that `blockexpiry` is being inserted into the arguments. But given that we'll either have an integer or a string, it should be possible to handle the older uses... Not sure I'm convinced though...

arcalinea replied:
Yeah, it complicates matters. The four options are:

1. No default, Command line arg. 
- Pros: Easily customizable by user, but not enforced. Cons
- Cons: Services have to upgrade to new format if they're expecting certain args.

2. No default, no command line arg: 
- Pros: Few worries for UI backwards compatibility
- Cons: Must construct custom transactions to use, which only Layer 2 protocols or services will likely be willing to do.

3. Default, command line arg:
- Pros: Most txs will have expiry, and is easily customizable from command line
- Cons: Requires the most work to make backwards compatible and integrate with services

4. Default, no command line arg:
- Pros: Many txs will have default expiry, and users or interfaces willing to construct custom transactions can customize.
- Cons: Perhaps too inflexible for users
 
Easiest to implement and upgrade to is 2. Could possibly add a default or command line arg later if it's well received?

str4d commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,61 @@
+
+Default behavior does not change -- transactions do not expire. (Alternative proposal is to set a default expiry period of 2 days, ~1440 blocks.) When nBlockExpiry is set to 0, the transaction will not expire from the mempool.
+
+If used in combination with nLockTime, nLockTime must be a blockheight, and nBlockExpiry must be higher. To simplify this feature, we could disallow its use in conjunction with nLockTime.
```
IIRC modern Bitcoin wallets set nLockTime to the most recent block to protect against some particular attack, so if we pulled that in (which we might not, or might do differently in order to sync it with the anchor selection), then users wouldn't be able to use `nBlockExpiry`.

We can just enforce in consensus that valid transactions have `nLockTime < nBlockExpiry` (and also correct relative units, if we add support for time-expiry too.

arcalinea replied:
The default behavior is to set nLockTime to current block - 10. I think it's fine to use in conjunction with nLockTime as long as both are blockheights; @bitcartel suggested disallowing its use with nLocktime.

str4d commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,61 @@
+Default behavior
+-----------------
+
+Default behavior does not change -- transactions do not expire. (Alternative proposal is to set a default expiry period of 2 days, ~1440 blocks.) When nBlockExpiry is set to 0, the transaction will not expire from the mempool.
```
Hmm... I'm kinda keen to follow the `nSequence` model here, where `UINT_MAX` is the default case, and anything lower is considered a valid block expiry. In particular, I'm thinking of how `nSequence` and `nLockTime` interact (but there isn't a split-range in `nSequence` like there is in `nLockTime`).

arcalinea replied:
If we make it `nExpiryTime` and introduce a split-range like in `nLockTime`, would this make a difference?  Higher numbers would be timestamps. And for `nLockTime`, 0 means no locktime, so it would be the same semantics for `nExpiryTime`. 

Another consideration is minimum expiration time. There should probably be a minimum expiration block height, which is something like chainActive.height + 720 (~1 day worth of blocks) to avoid users spamming the network with soon-to-be-expired txs. Some discussion on this PR rejected from Bitcoin for reorg issues: https://github.com/bitcoin/bitcoin/pull/3509
(Think we already face some of these reorg issues due to the way joinsplits are anchored, so can address the same way, by invalidating all dependent txs for a tx which gets dropped from chain upon reorg?)

arcalinea commented:
@str4d you convinced me on some of the points, changed to reflect your input and incorporated ideas from this rejected Bitcoin PR from 2014: https://github.com/bitcoin/bitcoin/pull/3509

arcalinea commented:
Expanding this inline comment that got squashed here:

The four options on how to implement this are:

 1.  No default, Command line arg.
    Pros: Easily customizable by user, but not enforced.
    Cons: Services have to upgrade to new format if they're expecting certain args.

 2. No default, no command line arg:
    Pros: Few worries for UI backwards compatibility
    Cons: Must construct custom transactions to use, which only Layer 2 protocols or services will likely be willing to do.

3. Default, command line arg:
    Pros: Most txs will have expiry, and is easily customizable from command line
    Cons: Requires the most work to make backwards compatible and integrate with services

4. Default, no command line arg:
    Pros: Many txs will have default expiry, and users or interfaces willing to construct custom transactions can customize.
    Cons: Perhaps too inflexible for users

Easiest to implement and upgrade to is 2. What do others think? Could possibly add a default or command line arg later if it's well received?

lindanlee commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,76 @@
+Wallet should notify user of expired transactions that must be re-sent.
+
+Wallet should notify user and reject the creation of a transaction that builds on a transaction with zero confirmations and an expiry blockheight set.
+
```
If a wallet predict which transactions are too large or insufficiently incentivized, I suggest they do that in a conservative manner. I know that this depends on the state of the network, though.

lindanlee commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,76 @@
+
+With the addition of this feature, zero-confirmation transactions with an expiration block height set will have even less guarantee of inclusion. This means that UIs and services must never rely on zero-confirmation transactions in Zcash.
+
+Wallet should notify user of expired transactions that must be re-sent.
```
In addition to this, can expired transactions be prevented? Error prevention, not resolution, is the best way to handle errors. 

If a wallet predict which transactions are too large or insufficiently incentivized, I suggest they do that in a conservative manner. I know that this depends on the state of the network and require some technical work for this to happen in a meaningful way without many false positives.

lindanlee commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,76 @@
+
+Transactions will have a new field, nExpiryTime, which will set the block height or UNIX timestamp after which transactions will be removed from the mempool if they have not been mined.
+
+The data type for nExpiryTime will be uint32_t, conforming to the structure of nLockTime. If used in combination with nLockTime, both nLockTime and nExpiryTime must be the same unit (either block height or timestamp).
```
I vote for using block height, since timestamp seems like a proxy for block height. If something expired, it's not because it turned a certain time, but because the blockchain reached a specific block height.

lindanlee commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,76 @@
+Wallet behavior and UI
+-----------------------
+
+With the addition of this feature, zero-confirmation transactions with an expiration block height set will have even less guarantee of inclusion. This means that UIs and services must never rely on zero-confirmation transactions in Zcash.
```
As a user researcher, I think it's good in the long term to have this be the standard. It's 100% worth it to prevent headache-inducing mistakes, like spending money you didn't have, on behalf of the user. 

I could see how some people might be frustrated if they were accustomed to using unconfirmed funds, or if funds take a long time to confirm (but then I suggest we solve the "it takes a while" problem and stick with not allowing users to spend unconfirmed funds).

bitcartel replied:
Today, it is not possible to zero conf spend the output notes of a shielded transaction, in another transaction.  Will this be possible in Sapling @sean @daira ?

bitcartel replied:
At the moment 0 conf transparent txs are fine, but 0 conf shielded tx are not.  It can be confusing to users.  However, I disagree that services should never rely on 0 conf txs just because the expiry height has been added.  There has never been a guarantee of inclusion for any tx so nothing has really changed i.e. a service should never assume that a 0 conf tx will eventually be mined.

daira replied:
@bitcartel No, that will not be possible in Sapling either.

daira replied:
Like @bitcartel, I disagree that services should (or must) never rely on 0-conf transactions. However, they shouldn't do so without looking at the expiration height, which could be very soon after they saw the transaction.

lindanlee commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,76 @@
+
+Wallet should notify user of expired transactions that must be re-sent.
+
+Wallet should notify user and reject the creation of a transaction that builds on a transaction with zero confirmations and an expiry blockheight set.
```
+1

lindanlee commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,76 @@
+1 hour before expiring, CreateNewBlock does not include the tx anymore. This is in case mining software changes the timestamp afterwards, resulting in the mining of an invalid block that contains expired transactions.
+1 hour after expiring, wallet marks the transaction as expired and frees the coins.
+
+Another reason to set a minimum expiration period is to mitigate the use of this feature as a spam mechanism.
```
I like the argument for setting a minimum block height to mitigate spam. But..  the user doesn't need to set the block height or to know it's the minimum, right? What is the threat model here?

lindanlee commented on drafts/arcalinea-tx-expiry/draft.rst:
````diff
@@ -0,0 +1,76 @@
+"version": 3,
+"locktime": 2089,
+"blockexpiry": 3539,
+```
````
Should we include what the block height was at the time the transaction was made? Maybe it's not necessary, but thought I would ask.

arcalinea replied:
I don't think it's necessary, is there a definite use case for this?

prestwich commented:
Looks pretty good to me. Any reason to require nLocktime and nExpiretime to use the same unit? I agree with the decision for cleanliness alone :)

Have you considered `OP_CHECKEXPIRETIMEVERY`? Implementation is similar to `OP_CLTV`. Can replace a NOP as a soft fork if tx expiry is implemented (must be paired with OP_DROP). Interestingly, expire time can _also_ be done as a soft fork by hacking it into sequence numbers and/or unused nLocktime bits. But don't, it's a mess. Summary: Script execution fails if any of these conditions are met:

1. TX version does not signal compatibility (as in CSV).
2. The stack is empty.
3. The top stack item is <0.
4. If nLocktime is non-0: nExpiretime < nLocktime
5. The nExpiretime is measured in blocks, but the top stack item is measured in seconds, or vice versa.
6. The top stack item is less than nExpiretime.

Has there been any consideration of relative transaction expiry? It would be pretty easy to implement alongside this (and fix everything about nSequence while you're at it). Zcash doesn't have CSV support yet afaik, so this might be a good time to add that in as well.

Useful reference for Bitcoin's existing timelock code: https://prestwi.ch/bitcoin-time-locks/

bitcartel commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,75 @@
+
+Transactions will have a new field, nBlockExpiry, which will set the block height after which transactions will be removed from the mempool if they have not been mined.
+
+The data type for nBlockExpiry will be uint32_t, conforming to the structure of nLockTime. If used in combination with nLockTime, both nLockTime and nBlockExpiry must be block heights.
```
"conforming to the structure" implies nBlockExpiry will also have the same semantic interpretation i.e. it can store a unix timestamp instead of just a block height. https://en.bitcoin.it/wiki/Protocol_documentation#tx

"If used in combination..." does this mean if nLockTime is ignored by the transaction, that nBlockExpiry can be a unix timestamp?

arcalinea replied:
No, can only be a block height. Will clarify this section

bitcartel commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,75 @@
+Default: TBD. One suggestion is 576 blocks, or about 1 day assuming 2.5 minute block times. Can add a config option to set user's default.
+Minimum: No minimum
+Maximum: 500000000, about 380 years
+No limit: To set no limit on transactions (so that they do not expire), nBlockExpiry should be set to UINT_MAX.  
```
UINT32_MAX or numeric_limits<uint32_t>::max() ... but since the code bases uses an int for the block height e.g. `(int)chainActive.Height()` this means that UINT_MAX as an int would represent a negative block height and thus cause many block height comparisons to fail.

arcalinea replied:
After conversation with Zooko on Rocketchat, and feedback from Linda, a proposal was made to change the default to 20 blocks.

arcalinea replied:
@daira what do you think the setting for "no expiry" should be?

bitcartel commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,75 @@
+Upon block reorganization
+--------------------------
+
+Transactions that are confirmed close to the end of their expiry period may be dropped from the mempool upon a block reorg, which could leave dependent transactions stranded in the mempool. Therefore, each time an expired transaction is removed from the mempool, a check must be added to remove its dependent transactions as well.
```
This check would be applied regardless of a block reorg.  That is, every time a transaction expires and should be removed from the mempool, so should all its dependent transactions.

bitcartel commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,75 @@
+
+Wallet should notify user of expired transactions that must be re-sent. See "Notify" section below.
+
+Wallet should notify user and reject the creation of a transaction that builds on a transaction with zero confirmations and an expiry blockheight set.
```
Since we intend the default behaviour to set a block expiry height on all txs, this would mean transparent tx become unusable.  Example, using 'sendtoaddress' to send funds to party A, and then immediately use 'sendtoaddress' to send funds to party B - but this call will fail since the first transaction has not been confirmed yet.

arcalinea replied:
So should we allow 0conf txs to be built upon?

bitcartel commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,75 @@
+Notify
+-------
+
+-expirenotify= can notify an external script when a wallet transaction expires
```
Executing scripts is hazardous, and we inherited -alertnotfiy.  Notification is down the implementation.  For the reference implementation zcashd, maybe zmq and amqp notifications are better?

arcalinea replied:
Ok, at any rate this is out of scope for now.

arcalinea commented:
@prestwich yes, cleanliness was why we're adding a new field and specifying only block height. Did not want to hack sequence numbers or unused bits.

OP_CSV or EXPIRETIMEVERIFY are out of scope for this feature, but if you'd like to see them included, it'd be great if you could open a ZIP or issue with the feature request.

arcalinea commented:
Think tx expiry should be user configurable, for following reasons: 
- will be easy for users to override if they dislike the feature
- will be easy to override if there is a bug
- will be easier to test

bitcartel commented:
@str4d @arcalinea To avoid type narrowing issues as previously discussed, how about:

- Let block expiry be a uint32_t
- 0 means no expiry.
- The maximum threshold for block expiry is 500000000 (same as lock time threshold, if > than this, tx is invalid) which fits comfortably in an `int32_t`

This would also allow us to one day support unix timestamps (even though we probably won't).

_Maybe already discussed but I want to make sure tx format PR is on the same page_

arcalinea commented:
sgtm, that makes sense. will change PR to reflect

bitcartel commented:
@arcalinea ZIP and PR should be modified for this fix regarding the threshold. https://github.com/zcash/zcash/pull/2925#discussion_r168822112

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -42,6 +42,7 @@ Default: TBD. Current proposal is 24 blocks, or about 1 hour assuming 2.5 minute
 Minimum: No minimum
 Maximum: 500000000, about 380 years
```
499999999

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,75 @@
+
+Transactions will have a new field, nExpiryHeight, which will set the block height after which transactions will be removed from the mempool if they have not been mined.
+
+The data type for nExpiryHeight will be uint32_t. If used in combination with nLockTime, both nLockTime and nExpiryHeight must be block heights. nExpiryHeight will never be a UNIX timestamp, unlike nLockTime values, and thus the maximum expiry height will be 500000000. 
```
Fixed.

daira replied:
s/500000000/499999999/

tromer commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,76 @@
+
+Advantages include optimizing mempool performance by removing transactions unlikely to be mined, and potentially simplifying bidirectional payment channels by reducing the need to store and compress revocations for past states, since transactions not committed to the chain could expire and become invalid after a period of time.
+
+If the expiry is block height N, it means the transaction must have either: 1. have been included in N-1, or 2. be included in block N. N+1 will be too late, and the transaction will be removed from the mempool.
```
"N+1 will be too late, and the transaction will be removed from the mempool" is the recommended behavior for nodes and miners, but the ZIP is still missing a clear specification of the *consensus rule*, like "the transaction will not be considered valid if included in block of depth greater than N".

arcalinea commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,75 @@
+
+Wallet should notify user of expired transactions that must be re-sent. See "Notify" section below.
+
+Wallet should notify user when building on a transaction with zero confirmations and an expiry blockheight set. Message TBD.
```
Should this exist as a UX feature? Deleting for now, but will open a ticket

arcalinea replied:
https://github.com/zcash/zcash/issues/3035

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,83 @@
+Abstract
+===========
+
+This is an Standards ZIP describing a new consensus rule to set an expiration time for transactions that are not mined to be removed from the mempool.
```
s/an/a/

daira replied:
"to set an expiration time after which a transaction cannot be mined. If it is not mined within that time, the transaction will be removed from nodes' mempools."

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,83 @@
+
+If the expiry is at block height N, then the transaction must be included in block N or earlier. Block N+1 will be too late, and the transaction will be removed from the mempool.
+
+The new consensus rule will enforce that the transaction will not be considered valid if included in block of height greater than N, and blocks that include transactions which have expired will not be considered valid.
```
s/transactions which have expired/expired transactions/

(See https://en.oxforddictionaries.com/usage/that-or-which . It would be correct to say "blocks that include transactions that have expired", but it sounds better to avoid the double "that".)

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,83 @@
+Specification
+===============
+
+Transactions will have a new field, nExpiryHeight, which will set the block height after which transactions will be removed from the mempool if they have not been mined.
```
```
``nExpiryHeight``
```

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,83 @@
+
+Transactions will have a new field, nExpiryHeight, which will set the block height after which transactions will be removed from the mempool if they have not been mined.
+
+The data type for nExpiryHeight will be uint32_t. If used in combination with nLockTime, both nLockTime and nExpiryHeight must be block heights. nExpiryHeight will never be a UNIX timestamp, unlike nLockTime values, and thus the maximum expiry height will be 500000000. 
```
```
``nExpiryHeight``
``uint32_t``
``nLocktime``
```

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
`````diff
@@ -0,0 +1,83 @@
+
+For the example below, the last block that the transaction below could possibly be included in is 3539. After that, it will be removed from the mempool.
+
+````
`````
three backticks, not four

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,83 @@
+No limit: To set no limit on transactions (so that they do not expire), nExpiryHeight should be set to 0.
+Coinbase: nExpiryHeight on coinbase transactions is ignored, and is set to 0 by convention.
+
+Every time a transaction expires and should be removed from the mempool, so should all its dependent transactions.
```
s/all its/all of its/

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
````diff
@@ -0,0 +1,83 @@
+"expiryheight": 3539,
+```
+
+Default: TBD. Current proposal is 24 blocks, or about 1 hour assuming 2.5 minute block times. Can add a config option to set user's default.
````
#2874 implements [a default of 20 blocks](https://github.com/zcash/zcash/pull/2874/files#diff-e8db9b851adc2422aadfffca88f14c91R72), and does have the configuration option.

arcalinea replied:
It does implement a configuration option, actually

daira commented on drafts/arcalinea-tx-expiry/draft.rst:
```diff
@@ -0,0 +1,83 @@
+
+With the addition of this feature, zero-confirmation transactions with an expiration block height set will have even less guarantee of inclusion. This means that UIs and services must never rely on zero-confirmation transactions in Zcash.
+
+Wallet should notify user of expired transactions that must be re-sent. See "Notify" section below.
```
"Wallets should notify the user"
